Alternative solution for pg_rewind missing wal

l-wang opened the pull request:
Implementing Ashwin's idea:

https://github.com/greenplum-db/gpdb/pull/14672#issuecomment-1362308762


## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

ashwinstar commented:
This code change looks and feels so much cleaner and easier. Still continuing to think on the impact.

Do we know if local `RedoRecPtr` is at all updated for walsender? and if yes at what frequency? (maybe can add log message to `GetRedoRecPtr()` and check without your changes). Based on code, it seems walsender won't ever refresh its local copy for `RedoRecPtr` as doesn't write WAL only reads it.

ashwinstar commented on src/backend/replication/walsender.c:
```diff
@@ -1772,13 +1772,14 @@ PhysicalConfirmReceivedLocation(XLogRecPtr lsn)
 	Assert(lsn != InvalidXLogRecPtr);
 	SpinLockAcquire(&slot->mutex);
-	if (slot->data.restart_lsn != lsn)
+	if (slot->data.restart_lsn < last_chkpt)
```
Doesn't this check will trigger unnecessary `changed = true;` setting if new checkpoints happen without any change to `lsn` reported by walreceiver?

l-wang replied:
You are correct! Fixing it.

l-wang commented:
> Do we know if local `RedoRecPtr` is at all updated for walsender? and if yes at what frequency? (maybe can add log message to `GetRedoRecPtr()` and check without your changes). Based on code, it seems walsender won't ever refresh its local copy for `RedoRecPtr` as doesn't write WAL only reads it.

You are correct. We can't rely on the walsender's local `RedoRrecPtr` because it doesn't get updated. walwriter would have it updated when it writes wals but walsender only updates its local `RedoRecPtr` at process start.

ashwinstar commented on src/backend/replication/walsender.c:
```diff
@@ -1766,19 +1766,38 @@ ProcessStandbyMessage(void)
+	 * one (internally created) replication slot, so this should not introduce
+	 * much contention on acquiring the spin lock.
+	 */
+	XLogRecPtr	last_chkpt = GetRedoRecPtr();
```
Just wish to note - `GetRedoRecPtr()` can return `InvalidXLogRecPtr`. Though based on how check is written `slot->data.restart_lsn < last_chkpt` seems don't require checking for invalid record pointer and current code should work fine.

ashwinstar reviewed (approved):
Change look good to me.

As part of this PR, we have test to validate WAL is not deleted for required cases. Please can you check if we have existing test to validate WAL is being deleted properly to the required point. If not maybe having such test would also be helpful to make sure no regression is being introduced from change, as it can go silently undetected and cause WAL space bloat in production systems.

l-wang commented:
> As part of this PR, we have test to validate WAL is not deleted for required cases. Please can you check if we have existing test to validate WAL is being deleted properly to the required point. If not maybe having such test would also be helpful to make sure no regression is being introduced from change, as it can go silently undetected and cause WAL space bloat in production systems.

Good call. We were missing coverage for WAL removal indeed. I added an isolation2 test to cover it. The test also reflects the changes we made for replication slots in this PR.

More context:
Upstream has one relevant tap test: [019_replslot_limit.pl](https://github.com/postgres/postgres/blob/master/src/test/recovery/t/019_replslot_limit.pl). It tests wal retention w.r.t. GUC `max_slot_wal_keep_size`, not exactly what we are looking for in this PR but some what along the same lines. 7X has an open issue https://github.com/greenplum-db/gpdb/issues/12007 for backporting this GUC. Previously we had added some wal related tests for 7X in commit https://github.com/greenplum-db/gpdb/commit/0668642213bf81f5db9d99e420af9caa74d5e407 and https://github.com/greenplum-db/gpdb/commit/06e7bf3269c5791edacbe90026e936cd22d64c2e , but neither covers WAL removal.
